### PR TITLE
Add feature gate JobTrackingWithFinalizers

### PR DESCRIPTION
**Pull request by alculquicondor**
And describe its behavior.

kubernetes/enhancements#2307

/sig apps

### CODE REVIEW

**Comment by alculquicondor** [comment]
/hold
for implementation to merge kubernetes/kubernetes#98238

/assign @soltysh

**Review by soltysh — APPROVED** [review]
/lgtm

File: content/en/docs/concepts/workloads/controllers/job.md
Comment (sftim) [inline content/en/docs/concepts/workloads/controllers/job.md:582]: I'd write:
```suggestion
When enabled, the control plane tracks new Jobs using the behavior described
```
(just a nit)

File: content/en/docs/concepts/workloads/controllers/job.md
Comment (sftim) [inline content/en/docs/concepts/workloads/controllers/job.md:586]:
```suggestion
When this feature isn't enabled, the Job {{< glossary_tooltip term_id="controller" >}} relies on counting the Pods
that exist in the cluster to
```

File: content/en/docs/concepts/workloads/controllers/job.md
Comment (sftim) [inline content/en/docs/concepts/workloads/controllers/job.md:592]:
```suggestion
- Human intervention to delete Pods belonging to a Job.
- An external controller (not provided as part of Kubernetes) that removes or replaces Pods.
```

File: content/en/docs/concepts/workloads/controllers/job.md
Comment (sftim) [inline content/en/docs/concepts/workloads/controllers/job.md:594]:
```suggestion
If you enable the `JobTrackingWithFinalizers` feature for your cluster, the control plane keeps track of
the Pods that belong to any Job and notices if any such Pod is
```

File: content/en/docs/concepts/workloads/controllers/job.md
Comment (sftim) [inline content/en/docs/concepts/workloads/controllers/job.md:596]: We'll soon have a glossary entry for finalizer, I hope. Mentioning this here as a tip for PR #28840

File: content/en/docs/concepts/workloads/controllers/job.md
Comment (sftim) [inline content/en/docs/concepts/workloads/controllers/job.md:601]:
```suggestion
before the feature is enabled are unaffected. You can determine if the Job
```

File: content/en/docs/concepts/workloads/controllers/job.md
Comment (sftim) [inline content/en/docs/concepts/workloads/controllers/job.md:603]:
```suggestion
`batch.kubernetes.io/job-tracking`. You should **not** manually add or remove this
```

**Review by sftim — COMMENTED** [review]
> The annotation will be removed no less than 2 releases after this feature graduates to GA, at which point all Jobs will be tracked using finalizers.

We need to mention this detail in 2 more places:
- https://kubernetes.io/docs/reference/labels-annotations-taints/
- https://kubernetes.io/docs/reference/using-api/deprecation-guide/ (not quite right but the best place to record it IMO)

**Review by alculquicondor — COMMENTED** [review]
> We need to mention this detail in 2 more places:
> 
> * https://kubernetes.io/docs/reference/labels-annotations-taints/

Added

> * https://kubernetes.io/docs/reference/using-api/deprecation-guide/ (not quite right but the best place to record it IMO)

Since we are not sure when we are graduating to GA, maybe we can hold on this for now. WDYT?

File: content/en/docs/concepts/workloads/controllers/job.md
Comment (shannonxtreme) [reply content/en/docs/concepts/workloads/controllers/job.md:600]: Replaced that PR with #28870 because I accidentally pushed the Hugo themes folder and...yeah :(

File: content/en/docs/concepts/workloads/controllers/job.md
Comment (sftim) [inline content/en/docs/concepts/workloads/controllers/job.md:610]: We usually [avoid statements about the future](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-statements-about-the-future) but this is a clear exception.



```suggestion
this annotation from Jobs.

#### Transition to stable feature {#job-tracking-with-finalizers-graduation}

The `JobTrackingWithFinalizers` feature will graduate to 
[general availability](/docs/reference/command-line-tools-reference/feature-gates/#feature-stages) (GA),
or might potentially be deprecated if testing reveals a design defect.
It's likely that this feature would reach general availability, and in that case then the
`batch.kubernetes.io/job-tracking` annotation will from that point remain honored
for at least 2 further minor Kubernetes releases. After _at least_
those 2 further releases, the `batch.kubernetes.io/job-tracking` annotation would become
deprecated and Kubernetes will switch to tracking completion for all Jobs using finalizers.
```

Comment (alculquicondor) [reply content/en/docs/concepts/workloads/controllers/job.md]: Done

File: content/en/docs/concepts/workloads/controllers/job.md
Comment (tengqm) [inline content/en/docs/concepts/workloads/controllers/job.md:607]: > You can determine if the Job controller is tracking a Job using finalizers if they have the annotation `batch.kubernetes.io/job-tracking`.

This sentence needs tweaking.

File: content/en/docs/concepts/workloads/controllers/job.md
Comment (tengqm) [inline content/en/docs/concepts/workloads/controllers/job.md:620]: This subsection is about future, so it can be safely removed.

File: content/en/docs/concepts/workloads/controllers/job.md
Comment (sftim) [inline content/en/docs/concepts/workloads/controllers/job.md:583]:
```suggestion
below. Existing Jobs are unaffected. As a user, the only difference you would see
is that the control plane tracking of Job completion is more accurate.
```
?

_(OK to merge without this change)_

**Review by sftim — COMMENTED** [review]
LGTM
@erictune / @soltysh you're reviewers for this page - do these changes look OK?

File: content/en/docs/concepts/workloads/controllers/job.md
Comment (alculquicondor) [reply content/en/docs/concepts/workloads/controllers/job.md]: Sg. Done

File: content/en/docs/concepts/workloads/controllers/job.md
Comment (alculquicondor) [reply content/en/docs/concepts/workloads/controllers/job.md:608]: better?

File: content/en/docs/concepts/workloads/controllers/job.md
Comment (alculquicondor) [reply content/en/docs/concepts/workloads/controllers/job.md]: Sg. Removed

**Comment by soltysh** [comment]
> @erictune / @soltysh you're reviewers for this page - do these changes look OK?

yes they do, I already left approval earlier :smile:

**Comment by alculquicondor** [comment]
rebased

File: content/en/docs/concepts/workloads/controllers/job.md
Comment (tengqm) [reply content/en/docs/concepts/workloads/controllers/job.md:601]: Sounds like you want to say this:

You can determine if the Job controller is tracking a Job using finalizers by checking if the Job has the annotation `batch.kubernetes.io/job-tracking`.

**Comment by tengqm** [comment]
@alculquicondor Looks like the gate has some problems. Please try trigger a new commit?

**Comment by alculquicondor** [comment]
It looks like I had a typo in `false`. Fixed :)

File: content/en/docs/concepts/workloads/controllers/job.md
Comment (tengqm) [reply content/en/docs/concepts/workloads/controllers/job.md:601]: ?

Comment (alculquicondor) [reply content/en/docs/concepts/workloads/controllers/job.md:601]: Better, thanks

**Comment by tengqm** [comment]
I approved this one because there have been previous approvals from both docs and tech side and the upstream PRs are all in (including the bug fixing one). I'm leaving the decision of lifting the hold to someone else for double confirmation.

**Pull request merged**
